integration/test/basic/main_test.go

```diff
@@ -84,17 +85,10 @@ func init() {
 func TestMain(m *testing.M) {
 	ctx := context.Background()
 
-	{
-		c := e2esetup.Config{
-			HelmClient: helmClient,
-			Host:       h,
-		}
-
-		v, err := e2esetup.Setup(ctx, m, c)
-		if err != nil {
-			l.LogCtx(ctx, "level", "error", "message", "e2e test setup failed", "stack", fmt.Sprintf("%#v\n", err))
-		}
-
-		os.Exit(v)
+	v, err := setup.WrapTestMain(h, helmClient, l, m)
+	if err != nil {
+		l.LogCtx(ctx, "level", "error", "message", "e2e test failed", "stack", fmt.Sprintf("%#v\n", err))
 	}
+
+	os.Exit(v)
```

Comment:
This integer dispatching and exit management is awful as hell. I cannot help but it just feels painful. How can we improve this? What is the actual problem we try to address with it?

Reply:
The issue we had here was that we only want to really run a test if the setup was successful, but still want to do the teardown even if it wasn't successful. And then we want to have errors from all stages but preserve the exit code if something along the way went wrong.

It can definitely be improved but I'd like to do that in a separate PR.